server/src/main/java/org/elasticsearch/cluster/coordination/ClusterStatePublisher.java

```diff
@@ -8,36 +8,50 @@
 package org.elasticsearch.cluster.coordination;
 
 import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.ClusterStateApplier;
 import org.elasticsearch.cluster.ClusterStatePublicationEvent;
+import org.elasticsearch.cluster.Diff;
 import org.elasticsearch.cluster.node.DiscoveryNode;
 import org.elasticsearch.core.Nullable;
 import org.elasticsearch.core.TimeValue;
 
 public interface ClusterStatePublisher {
     /**
-     * Publish all the changes to the cluster from the master (can be called just by the master). The publish
-     * process should apply this state to the master as well!
+     * Publishes an updated {@link ClusterState} to all the nodes in the cluster:
      *
-     * The publishListener allows to wait for the publication to complete, which can be either successful completion, timing out or failing.
-     * The method is guaranteed to pass back a {@link FailedToCommitClusterStateException} to the publishListener if the change is not
-     * committed and should be rejected. Any other exception signals that something bad happened but the change is committed.
+     * <ul>
+     *     <li>Send the updated state to all nodes. Typically we only send a {@link Diff} which each node combines with its local state to
+     *     yield the updated state very cheaply.</li>
+     *     <li>Wait for enough master-eligible nodes to indicate that they have <i>accepted</i> the new state (i.e. written it to durable
+     *     storage).</li>
+     *     <li>Tell all the nodes that the new state is now committed and should be applied via {@link
+     *     ClusterStateApplier#applyClusterState}.</li>
+     *     <li>Finally, apply the state on the master.</li>
+     * </ul>
      *
-     * The {@link AckListener} allows to keep track of the ack received from nodes, and verify whether
-     * they updated their own cluster state or not.
+     * @param publishListener notified when the publication completes, whether successful or not. If successful then the new state has been
+     *                        applied on this node and usually all other nodes (although some of them might have failed or timed out
+     *                        instead). May fail with a {@link FailedToCommitClusterStateException} if this node didn't hear that enough
+     *                        other master-eligible nodes accepted this state, which causes this node to stand down as master. On failure
+     *                        the new state might or might not take effect in some future update: if it was accepted by the new master then
```

Comment:
I am not sure I follow "the new state might or might not take effect in some future update" , i think that wording implies that we could drop an update on floor ?

Reply:
Yes that can happen if a publication fails (but it doesn't necessarily happen). I expanded and reordered the words here, hope they're better now.

Reply:
Perfect !  Very helpful clarification.